pkg/pubsub/client/pubsub_client.go

```diff
@@ -0,0 +1,93 @@
+package client
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/dapr"
+)
+
+func init() {
+	flag.Var(pubSubs, "pub-sub-tool", "Preferred pub-sub tool, e.g dapr, rabbitmq. This flag can be declared more than once. Omitting will default to not using any pub sub tool.")
+}
+
+var pubSubs = newPubSubSet(map[string]InitiateClient{
+	dapr.Name: dapr.NewClient,
+},
+)
+
+// PubSub is the interface that wraps pubsub methods.
+type PubSub interface {
+	// Publish single message over a specific topic/channel
+	Publish(data interface{}, topic string) error
+
+	// Publish a batch of messages over a specific topic/channel
+	PublishBatch(data interface{}, topic string) error
+
+	// Get the name of pub sub tool
+	GetName() string
+
+	// Determines if publish batching is enabled or not
+	IsBatchingEnabled() bool
+}
+
+type pubSubSet struct {
+	supportedPubSub map[string]InitiateClient
+	enabledPubSub   map[string]InitiateClient
+}
+
+type InitiateClient func(context.Context, string) (interface{}, error)
```

Comment:
How are we passing config to the client?

If we require the client to gather the config itself, we run the risk of clients being coupled to particular environments (it looks like Dapr may require an active cluster, which is an issue, IMO, particularly for `gator test`).

If we allow callers to pass in a config, it's up to the caller how that config is discovered (e.g. `gator test` loading it from the config stream or -- to prevent credential leakage -- a separate config directory)

Reply:
can you elaborate more? I am not following your thoughts here

Reply:
> If we allow callers to pass in a config, it's up to the caller how that config is discovered (e.g. gator test loading it from the config stream or -- to prevent credential leakage -- a separate config directory)

The 2nd parameter passed in this function is expected to be a config. So whoever is calling to `InitiateClient` is responsible for generating valid config.

Reply:
I see, so config is a string? it may be worth naming the variables for clarity.

Do we expect config will be a string? `map[string]interface{}` if we are reading from config maps?

Reply:
Right now I am envisioning it to be a json map. For example,
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: dapr
  namespace: gatekeeper-system
data:
  connection1: |
    {
      "enableBatching": false,  
      "component": "pubsub",
      "size": 10
    }
  connection2: |
    {
      "enableBatching": false,  
      "component": "pubsub",
      "size": 10
    }
   ```
   
   +1 to adding variable name